readme.txt
=== Plugin Name ===
Contributors: igrossiter
Donate link: http://onespace.com.br/doacoes/
Tags: latest posts, post
Requires at least: 3.0.1
Tested up to: 3.9.2
Stable tag: trunk
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html

Provides a shortcode to get the latest posts on the blog, for one or more categories or a post_type.

== Description ==

Veja descrição em portugues em: http://onespace.com.br/lastpost/

This plugin provides you with a shortcode that returns the latest posts from specifics categories or post_types, you can choose how many posts is to be show or what kind of information you can get from posts.

Arguments:

1. id - You can use any number of post categories ids separeted by commas
1. num - the number of posts to return for wich categorie id
1. info - this is where goes all the printing information

Info printing arguments:

1. thumb: returns the post's featured image, you must specify the width of the thumb after ':', has a link to the post
1. title: returns the title, has a link to the post
1. excerpt: returns the first lines from you post you can specify the lenght of the text to show and a subtext. excerpt:LENGTH:SUB-TEXT, has a link to the post
1. content: returns all content from post
1. date: returns date of post publishing, you can specify the format after ':'

More about date:

Date format uses php default formating symbols, here goes some of then;

d: day of month
m: month
y: year

it has some limitations, you can't use escape chars (yet).

check the manual to see the complete list: http://php.net/manual/en/function.date.php

A few examples:

[os_lastpost id="x,y,z" num=2 info="title,date:d/m/y,thumb:130,excerpt:120: ... veja mais"]

returns the latest 2 posts from post categories x, y and z, printing the date on Day/Month/Year, featured image 130 px of width, 120 chars from post content (html stripped) and a custom text " ... veja mais" at the end.

[os_lastpost post_type="custom_post_type" num=5 info="title,excerpt"]

returns the latest 5 posts from a custom post type, printing title, and the first 60 chars from the post content and a default text " ... read more"

[os_lastpost post_type="post" num=5 info="title,content"]

returns the latest 5 posts from all your categories, printing title, and full content

== Installation ==

This section describes how to install the plugin and get it working.

e.g.

1. Unzip and upload `os-lastpost` to the `/wp-content/plugins/` directory
1. Activate the plugin through the 'Plugins' menu in WordPress
1. Use the shortcode described on the description in your site
1. If you need to use the shortcode on a widget, it is possible with a additional plugin

== Frequently Asked Questions ==

= I need a list on a widget =

Get a plugin that allows you to use any shortcode on widgets, thats fairly easy to find here on wp plugin repository.

= How do I found out the ID from a category? =

Go to Admin > posts > categories, mouse over the category that you need, and check the ID on the link.

== Screenshots ==

1. Plugin used on a post.

2. Shortcode used.

3. Plugin used on a widget, need additional plugin to make possible the use of shortcodes on widget.

== Changelog ==

= 0.1 =
inicial version

== Upgrade Notice ==

== Arbitrary section ==


changes.txt
- Info excerpt now respects the <!--more--> tag, content always show the full content


load.php
<?php
/* OneSpace Last Post
 * Version: 0.1
 * Author: Ian Garcez
 * Author URI: http://onespace.com.br
 * License: 
 * 
 * Copyright 2013  Ian Garcez  (email : ian@onespace.com.br)
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as 
 * published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

include_once 'core/class-os-lastpost.php';
include_once 'core/class-os-lastpost-register.php';
include_once 'core/class-os-lastpost-printer.php';
include_once 'plugin-functions.php';

os-lastpost.php
<?php
/*
Plugin Name: OneSpace Last Post
Plugin URI: http://onespace.com.br/lastpost
Description: Shortcode to get latest post with criteria.
Version: 0.1
Author: Ian Garcez
Author URI: http://onespace.com.br
License: 

    Copyright 2013  Ian Garcez  (email : ian@onespace.com.br)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License, version 2, as 
    published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/
include_once 'load.php';

$obj = new Os_lastpost;
$obj->register_shortcode();

plugin-functions.php
<?php
/* OneSpace Last Post
 * Version: 0.1
 * Author: Ian Garcez
 * Author URI: http://onespace.com.br
 * License: 
 * 
 * Copyright 2013  Ian Garcez  (email : ian@onespace.com.br)
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as 
 * published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

//init
add_action( 'wp_enqueue_scripts', 'os_lastpost_style' );
function os_lastpost_style() {
    wp_register_style( 'os-lastpost-style', plugins_url('style.css', __FILE__) );
    wp_enqueue_style( 'os-lastpost-style' );
}

core/class-os-lastpost.php
<?php
/* OneSpace Last Post
 * Version: 0.1
 * Author: Ian Garcez
 * Author URI: http://onespace.com.br
 * License: 
 * 
 * Copyright 2013  Ian Garcez  (email : ian@onespace.com.br)
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as 
 * published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

class Os_lastpost{
	// objeto para realizar as funções
	public $register;

	public function __construct() {
       $this->register = new Os_lastpost_register;
   }

	public function register_shortcode(){
		$this->register->register();
	}


}

core/class-os-lastpost-register.php
<?php
/* OneSpace Last Post
 * Version: 0.1
 * Author: Ian Garcez
 * Author URI: http://onespace.com.br
 * License:
 *
 * Copyright 2013  Ian Garcez  (email : ian@onespace.com.br)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

class Os_lastpost_register {
	public function register(){
		add_shortcode('os_lastpost', array( __CLASS__, 'os_lastpost_main' ));
	}

	public static function os_lastpost_main( $atts ){
		//retorna erro se id ou post_type não estiverem definidos
		if(!isset($atts['id']) && !isset($atts['post_type']))
			return '<span class="os_lastpost_error">nada a buscar</span>';

		// não busquermos id caso tenhamos um post_type definido
		if (isset($atts['post_type']) && $atts['post_type']!='post'){
			unset($atts['id']);
		}

		if(!isset($atts['post_type'])){
			$atts['post_type'] = 'post';
		}

		$printer = new Os_lastpost_printer($atts);

		return $printer->getHtml();

	}
}

core/class-os-lastpost-printer.php
<?php
/* OneSpace Last Post
 * Version: 0.1
 * Author: Ian Garcez
 * Author URI: http://onespace.com.br
 * License:
 *
 * Copyright 2013  Ian Garcez  (email : ian@onespace.com.br)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
class Os_lastpost_printer{
	public $html = '';

	private $querys_for_print = array();

	private $args = array(
		'id' => array(),
		'post_type' => 'post',
		'num' => 3,
		);

	private $info_to_print = array('title');

	// o construtor vai montar o html
	public function __construct($atts) {
		$this->makeArgs($atts);
		$this->makeQuerys($this->args);
		$this->makeHtml();
	}

	private function makeArgs($atts){
		if( $atts['id'] ){
			$list_ids = explode(',', $atts['id']);
			$this->args['id'] = $list_ids;
		}

		if( $atts['post_type'] != 'post'){
			$this->args['post_type'] = $atts['post_type'];
		}

		if( $atts['info'] ){
			$this->info_to_print = explode(',', $atts['info']);
		}

		if( $atts['num'] ){
			$this->args['num'] = $atts['num'];
		}
	}

	private function makeQuerys($a){
		if (!$a['id'] && $a['post_type'] == 'post'){
			$this->makeQueryForPostType($a);
		} else if($a['post_type'] != 'post') {
			$this->makeQueryForPostType($a);
		} else {
			$this->makeQueryForCats($a);
		}

	}

	private function makeQueryForPostType($a){
		$query_args = array(
			'post_type' => $a['post_type'],
			'posts_per_page' => $a['num'],
			);

		$this->querys_for_print[] = new WP_Query($query_args);

	}

	private function makeQueryForCats($a){
		foreach ($a['id'] as $cat) {
			$query_args = array(
				'category__in' => $cat,
				'post_type' => 'post',
				'posts_per_page' => $a['num'],
				);
			$this->querys_for_print[] = new WP_Query($query_args);
		}


	}

	private function makeHtml(){
		$this->html = '<div class="os-lastpost">';

		// save the current post for recovery later
		global $post;
		global $more;
		$save_post = $post;
		$save_more = $more;


		foreach ($this->querys_for_print as $q) {
			$this->html .= '<ul>';
			if ($q->have_posts()): while($q->have_posts()):	$q->the_post();
					$post = $q->post;
					$more = 0; // respect the <!--more--> tag
					$this->html .= '<li>';
					$this->makeHtmlWithInfo();
					$this->html .= '</li>';
				endwhile;
			endif;
			$this->html .= '</ul>';
		}
		$this->html .= '</div>';

		$post = $save_post;
		$more = $save_more;
	}

	private function makeHtmlWithInfo() {
		foreach ($this->info_to_print as $info) {
			if(strpos($info, 'title') !== false){
				$this->html .= '<h2 class="title"><a href="' . get_permalink() . '">';
				$this->html .= get_the_title().'</a></h2>';
			}

			if(strpos($info, 'date') !== false){
				$format = $this->getDateFormat($info);
				$this->html .= '<p class="date">'. get_the_time($format) . '</p>';
			}

			if(strpos($info, 'excerpt') !== false){
				$this->html .= $this->getExcerptFromInfo($info);
			}

			if(strpos($info, 'content') !== false){
				$this->html .= $this->getFullContentForInfo($info);
			}

			if(strpos($info, 'thumb') !== false){
				$this->html .= $this->getThumbsForInfo($info);
			}
		}
	}

	private function getDateFormat($string) {
		$string = explode(':', $string);
		if (count($string) > 1){
			return $string[1];
		} else {
			return 'm/d/y';
		}
	}

	private function getExcerptFromInfo($string) {
		$options = explode(':', $string);
		if (count($options) > 2){
			$max_char = intval($options[1]);
			$text = ($options[2]);
			return $this->trimContent($max_char, $text);

		} else if (count($options) > 1){
			$max_char = intval($options[1]);
			return $this->trimContent($max_char);

		} else {
			return $this->trimContent();
		}

	}

	private function trimContent($max_char = 60, $text = ' ...Read more') {
		$excerpt = get_the_content($text);
		$excerpt = strip_tags($excerpt);
		if (strlen($excerpt) > $max_char - strlen($text)){
			$excerpt = substr($excerpt, 0, $max_char) . $text;
		}


		return '<p><a href="' . get_permalink() . '">' . $excerpt . '</a></p>';
	}

	private function getFullContentForInfo($string){
		global $more;
		$more = 1; // alwayes show full content
		$content = get_the_content();
		$content = apply_filters('the_content', $content);
		return $content;
	}

	private function getThumbsForInfo($string){
		$string = explode(':', $string);
		$size = intval($string[1]);
		$size = array($size, $size);
		if ($size > 10){
			return '<a href="' . get_permalink() . '">' .get_the_post_thumbnail(get_the_ID(), $size ) . '</a>';
		} else {
			return $this->printError("Tamanho da imagem inválido");
		}
	}

	private function printError($string){
		return '<div class="os_lastpost_error">' . $string . '</div>';
	}

	public function getHtml(){
		return $this->html;
	}
}